docs/diagrams/catchup-procedure.puml

```diff
@@ -0,0 +1,317 @@
+                  <i>in this reply to last transaction to catch up)
+        end note
+
+        loop On each received catch-up reply
+            opt Solid range adjacent to ledger of not yet applied\nreceived catch-up replies exists
+                loop For each catch-up reply in solid range adjacent to ledger
+
+                    Node4 -> Node4: Verify catch-up reply
+                    note right Node4
+                        <b>Verify catch-up reply:
+                        Transactions and consistency proof in catch-up reply
+                        must conform with quorumed CONSISTENCY_PROOF
+                    end note
+
+                    Node4 -> Node4: Apply catch-up reply
+                    note right Node4
+                        <b>Apply catch-up reply:
+                        For each transaction in catch-up reply:
+                            - Add transaction to ledger in committed mode
+                            - Apply transaction to state in committed mode
+                    end note
+
+                end loop
+            end opt
+        end loop
+
+        note over Node1, Node4
+            Ledger reached target size
+        end note
+
+        Node4 -> Node4: Update txn_seq_range_to_3phase_key
+        note right Node4
+            Add <b>Master's 3PC key of last caught up transaction</b>
+            (gotten from <b>quorumed CONSISTENCY_PROOF</b>)
+            to <b>txn_seq_range_to_3phase_key</b>
+        end note
+
+        opt Ledger is pool
+            Node4 -> Node4: mode = discovered
+        end opt
+
+    end alt
+
+end loop
+
+Node4 -> Node4: Get last_caught_up_3PC
+note right Node4
+    Get <b>last_caught_up_3PC</b> as maximum of Master's 3PC keys
+    of last transactions over all ledgers
+end note
+
+opt last_caught_up_3PC > Master's last_ordered_3pc
+    loop For each replica
+        alt Replica is master
+
+            Node4 -> Node4: Replica's last_ordered_3pc = last_caught_up_3PC
+
+            Node4 -> Node4: Clear replica's collections up to last_caught_up_3PC
+            note right Node4
+                <b>Clear following replica's collections:
+                    - batches <i>- up to last_caught_up_3PC,
+                    - sentPrePrepares <i>- up to last_caught_up_3PC,
+                    - prePrepares <i>- up to last_caught_up_3PC,
+                    - prepares <i>- up to last_caught_up_3PC,
+                    - commits <i>- up to last_caught_up_3PC,
+                    - requestQueues <i>- requests from 3PC batches up to last_caught_up_3PC,
+                    - ORDEREDs from outBox  <i>- up to last_caught_up_3PC,
+                    - checkpoints  <i>- completely,
+                    - stashedRecvdCheckpoints  <i>- up to last_caught_up_3PC
+            end note
+
+            Node4 -> Node4: Update replica's watermarks to\n(last_caught_up_3PC[1], last_caught_up_3PC[1] + LOG_SIZE)
+
+        else Replica is backup
```

Comment:
Please mention (maybe in a form of TBD item?) why we are doing it differently on backups